File: py38/sort/select.py
from utils import timer

from typing import List


@timer
def select_sort(l: List):
    for i in range(len(l)):
        min_num = l[i]
        min_index = i
        for index, num in enumerate(l[i+1:]):
            if num < min_num:
                min_index = index+i+1
                min_num = num

        l[min_index], l[i] = l[i], l[min_index]


if __name__ == "__main__":
    l = [2, 1, 3, 4]
    select_sort(l)
    assert l == list(range(1, 5))

File: rs/src/sort/select.rs
pub fn select_sort<T: Ord>(l: &mut [T]) {
    for i in 0..l.len() {
        let mut min_index = i;
        let mut min_num = &l[i];

        for j in (i+1)..l.len() {
            if &l[j] < min_num {
                min_index = j;
                min_num = &l[j];
            }
        }

        l.swap(min_index, i);
    }
}